lib/prestashopper/api.rb

```diff
@@ -22,33 +22,65 @@ def initialize(url, key)
     # List resources that the API key can access
     # @return [Array<Symbol>] list of resources the API can access
     def resources
-      xml = @resources_res.get.body
-      xml_doc = Nokogiri::XML xml
-      nodes = xml_doc.xpath '/prestashop/api/*'
-      resources_list = []
-      nodes.each{|n| resources_list << n.name.to_sym}
-      return resources_list
-    end
-
-    # Get all products data
-    # @return [Array<Hash>] list of products. Each product is represented by a hash with all its attributes.
-    def get_products
-      # /api/products returns XML with the IDs of each individual product
-      xml_products = @resources_res['products'].get.body
-      xml_products_doc = Nokogiri::XML xml_products
-      products_nodes = xml_products_doc.xpath '/prestashop/products/*/@id'
-      ids_list = []
-      products_nodes.each{|n| ids_list << n.value}
-
-      # GET each individual product to get the whole data
-      products = []
-      ids_list.each do |id|
-        xml_product = @resources_res["products/#{id}"].get.body
-        product = Product.xml2hash xml_product
-        products << product
+      @resources ||= Nokogiri::XML(@resources_res.get.body).xpath('/prestashop/api/*').collect { |resource| resource.name.to_sym }
+    end
+
+    def method_missing(method, *args, &block)
+      if method.to_s.match(/^get_/)
+        resource = method.to_s.sub(/^get_/,'').pluralize.to_sym
+        raise RestClient::MethodNotAllowed, 'You do not have access to this resource' unless resources.include?(resource)
+        
+        if method.to_s == method.to_s.singularize
+          get_resource(resource, collect_ids_for_resource(args), true)
+        else
+          ids = collect_ids_for_resources(args)
+          ids.present? ? get_resources(resource, ids) : get_resource_ids(resource)
+        end
+      else
+        super
+      end
+    end
+
+    private
+
+    def get_resource_ids(resource)
+      Nokogiri::XML(@resources_res[resource].get.body).xpath("/prestashop/#{resource}/*/@id").collect(&:value)
+    end
+
+    def get_resource(resource, id, raise_not_found_exception = false)
+      resource_class_name = resource.to_s.classify
+      resource_class = "Prestashopper::#{resource_class_name}".safe_constantize || Prestashopper.const_set(resource_class_name, Class.new(OpenStruct))
```

Comment:
resource classes are generated on the fly, this will prevent the gem from needing an update when available resources are added.

Additionally, this code will still allow for a the class to be defined in case a custom parser is required or additional methods are added.